pkg/util/watch/watch.go

```diff
@@ -0,0 +1,100 @@
+package watch
+
+import (
+	"errors"
+	"time"
+
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+type WatcherFunc func(string) (watch.Interface, error)
+
+// RetryWatchUntil starts a watch and handle the watch events using the conditionFunc. In
+// case the watch is closed, this function will re-open it and continue handling events
+// from the last resourceVersion received.
+// The timeout can be set to limit the amount of time the conditionFunc should report
+// results.
+func RetryWatchUntil(watcher WatcherFunc, conditionFunc watch.ConditionFunc, timeout time.Duration) (bool, error) {
+	type watchResult struct {
+		err    error
+		result bool
+	}
+	eventChan := make(chan watch.Event)
+	stopChan := make(chan struct{})
+	resultChan := make(chan watchResult)
+	defer close(resultChan)
```

Comment:
Channels are garbage-collected once they go out of scope. Closing a channel means you want to broadcast. In your code, threads **send** to `resultChan` instead of receiving (so you don't need to broadcast) and moreover sending to a closed channel is a runtime error.